File: CV_2-TRANSFORMATIONS/exercise8.py
import cv2
import numpy as np


def create_beach_ball_animation():
    # Read images
    ball = cv2.imread('ball.jpg', cv2.IMREAD_UNCHANGED)
    # Read and invert the mask
    mask = cv2.imread('ball_mask.jpg', cv2.IMREAD_GRAYSCALE)
    mask = cv2.bitwise_not(mask)  # Invert the mask so ball is white and background is black
    background = cv2.imread('beach.jpg')

    # Resize background to a standard size
    background = cv2.resize(background, (1280, 720))

    # Resize ball and mask to be proportional to background
    ball_size = 100
    ball = cv2.resize(ball, (ball_size, ball_size))
    mask = cv2.resize(mask, (ball_size, ball_size))

    # Get dimensions
    height, width = background.shape[:2]

    # Create video writer
    fourcc = cv2.VideoWriter_fourcc(*'XVID')
    out = cv2.VideoWriter('transf_beach.avi', fourcc, 30.0, (width, height))

    # Animation parameters
    num_frames = 150

    # Ball movement parameters
    start_x = width // 4
    start_y = height // 2

    # Define ball path (arc motion)
    t = np.linspace(0, 1, num_frames)
    x_positions = np.linspace(start_x, width * 3 // 4, num_frames)
    y_positions = start_y - 200 * np.sin(np.pi * t)  # Create arc motion

    # Scale factors (ball gets smaller as it moves)
    scale_factors = np.linspace(1.0, 0.3, num_frames)

    # Rotation angles
    rotation_angles = np.linspace(0, 360, num_frames)

    for frame in range(num_frames):
        # Get current position and scale
        current_x = int(x_positions[frame])
        current_y = int(y_positions[frame])
        scale = scale_factors[frame]
        angle = rotation_angles[frame]

        # Scale ball and mask
        current_size = int(ball_size * scale)
        if current_size < 1:  # Prevent size from becoming too small
            current_size = 1
        scaled_ball = cv2.resize(ball, (current_size, current_size))
        scaled_mask = cv2.resize(mask, (current_size, current_size))

        # Create rotation matrix
        M = cv2.getRotationMatrix2D((current_size // 2, current_size // 2), angle, 1.0)

        # Rotate ball and mask
        rotated_ball = cv2.warpAffine(scaled_ball, M, (current_size, current_size))
        rotated_mask = cv2.warpAffine(scaled_mask, M, (current_size, current_size))

        # Create frame
        frame = background.copy()

        # Calculate ROI coordinates
        y1 = current_y - current_size // 2
        y2 = y1 + current_size
        x1 = current_x - current_size // 2
        x2 = x1 + current_size

        # Ensure coordinates are within frame bounds
        if x1 < 0: x1 = 0
        if y1 < 0: y1 = 0
        if x2 > width: x2 = width
        if y2 > height: y2 = height

        # Adjust ball and mask size if needed
        ball_roi = rotated_ball[:(y2 - y1), :(x2 - x1)]
        mask_roi = rotated_mask[:(y2 - y1), :(x2 - x1)]

        # Create ROI in the frame
        roi = frame[y1:y2, x1:x2]

        # Apply mask
        mask_roi = mask_roi > 128  # Create binary mask
        roi[mask_roi] = ball_roi[mask_roi]

        # Put ROI back into frame
        frame[y1:y2, x1:x2] = roi

        # Write frame
        out.write(frame)

    # Release video writer
    out.release()


try:
    create_beach_ball_animation()
    print("Beach ball animation created successfully!")
except Exception as e:
    print(f"An error occurred: {str(e)}")

File: CV_2-TRANSFORMATIONS/exercise8.m
function create_beach_ball_horizon_animation()
    try
        % Read images
        ball = imread('ball.jpg');
        mask = imread('ball_mask.jpg');
        background = imread('beach.jpg');

        % Convert images to double for processing
        ball = im2double(ball);
        mask = im2double(rgb2gray(mask));
        background = im2double(background);

        % Invert the mask
        mask = ~mask;

        % Resize background to standard size
        background = imresize(background, [720 1280]);

        % Initial ball size and resize
        ball_size = 100;
        ball = imresize(ball, [ball_size ball_size]);
        mask = imresize(mask, [ball_size ball_size]);

        % Get dimensions
        [height, width, ~] = size(background);

        % Create video writer
        v = VideoWriter('transf_beach.avi');
        v.FrameRate = 30;
        open(v);

        % Animation parameters
        num_frames = 150;

        % Create time vector
        t = linspace(0, 1, num_frames);

        % Starting positions
        start_x = width / 4;
        start_y = height / 2;

        % Calculate trajectories
        x_positions = linspace(start_x, width * 3/4, num_frames);
        y_positions = start_y - 200 * sin(pi * t);

        % Scale factors (ball gets smaller as it moves)
        scale_factors = linspace(1.0, 0.3, num_frames);

        % Rotation angles
        rotation_angles = linspace(0, 360, num_frames);

        % Main animation loop
        for frame = 1:num_frames
            % Get current position and parameters
            current_x = round(x_positions(frame));
            current_y = round(y_positions(frame));
            scale = scale_factors(frame);
            angle = rotation_angles(frame);

            % Scale ball and mask
            current_size = max(1, round(ball_size * scale));
            scaled_ball = imresize(ball, [current_size current_size]);
            scaled_mask = imresize(mask, [current_size current_size]);

            % Rotate ball and mask using bilinear interpolation
            rotated_ball = imrotate(scaled_ball, angle, 'bilinear', 'crop');
            rotated_mask = imrotate(scaled_mask, angle, 'bilinear', 'crop');

            % Create frame
            current_frame = background;

            % Calculate ROI coordinates
            y1 = max(1, round(current_y - current_size/2));
            y2 = min(height, y1 + current_size - 1);
            x1 = max(1, round(current_x - current_size/2));
            x2 = min(width, x1 + current_size - 1);

            % Calculate valid regions
            roi_height = y2 - y1 + 1;
            roi_width = x2 - x1 + 1;

            % Get regions from rotated images
            ball_roi = rotated_ball(1:roi_height, 1:roi_width, :);
            mask_roi = rotated_mask(1:roi_height, 1:roi_width);

            % Get ROI from current frame
            roi = current_frame(y1:y2, x1:x2, :);

            % Apply mask to each channel
            for c = 1:3
                roi(:,:,c) = ball_roi(:,:,c) .* mask_roi + roi(:,:,c) .* (1-mask_roi);
            end

            % Insert ROI back into frame
            current_frame(y1:y2, x1:x2, :) = roi;

            % Convert to uint8 before writing
            current_frame = im2uint8(current_frame);

            % Write frame
            writeVideo(v, current_frame);
        end

        % Close video writer
        close(v);
        disp('Beach ball horizon animation created successfully!');

    catch ME
        disp(['An error occurred: ' ME.message]);
        if exist('v', 'var') && isopen(v)
            close(v);
        end
    end
end

% Run the function
create_beach_ball_horizon_animation();